llvm/lib/ProfileData/InstrProf.cpp

```diff
@@ -300,12 +316,8 @@ getIRPGONameForGlobalObject(const GlobalObject &GO,
                             GlobalValue::LinkageTypes Linkage,
                             StringRef FileName) {
   SmallString<64> Name;
-  if (llvm::GlobalValue::isLocalLinkage(Linkage)) {
-    Name.append(FileName.empty() ? "<unknown>" : FileName);
-    Name.append(";");
-  }
   Mangler().getNameWithPrefix(Name, &GO, /*CannotUsePrivateLabel=*/true);
```

Comment:
Can we move the Mangler() handling into getGlobalIdentifier(), then remove the handling there of "Name[0] == '\1'" as I believe it is handled by Manger().getNameWithPrefix().

Reply:
@ellishg How much of D156569 relied on the invocation of Mangler? It is not mentioned in the patch description, only the rationale for changing ":" to ";". The problem is if these are out of sync, then cross-module importing of indirectly called local functions will continue to be broken in whatever cases Mangler().getNameWithPrefix affects.

Reply:
D156569 changes the format from `[<filepath>:]<function-name>` to `[<filepath>;]<linkage-name>`. Note the change from **function-name** to **linkage-name**. Having the mangled name is required so that we can pass a symbol order via `-symbol-ordering-file` or `-order_file`.

Reply:
Well, I guess the name is still mangled, but I used `Mangler().getNameWithPrefix()` to get the linkage name.

Reply:
> I guess the name is still mangled, but I used Mangler().getNameWithPrefix() to get the linkage name.

When raw profiles stores (optionally filename prefixed) linkage names while global identifier uses the (filename prefixed) mangled name, the particular use case would be broken. Is it possible to convert objective-c mangled name to linkage name (needed by linker) in the post-processing of raw profiles? And do you think it make sense for the use case of D156569?

Say raw profiles still stores `[filepath;]mangled-name`, and `llvm-profdata` takes it to generate the linkage name that `-symbol-ordering-file` wants.

Reply:
I think we can only generate the linkage name at compile time because we need to know the linkage, which `llvm-profdata` doesn't have.

Reply:
Can you give an example of where this is called such that there is a difference between the function name and linkage name (at least for c++ these are the same afaik), and what that difference looks like? Are there specific callsites used for the symbol ordering support that need the mangling?

Reply:
Here is the unittest for these functions which shows some examples. Even for c++ names there are differences

https://github.com/llvm/llvm-project/blob/70187ebadf20f014a7821cf322eae60192dbe4cc/llvm/unittests/ProfileData/InstrProfTest.cpp#L542-L609

Reply:
> Was PGO broken in general before D156569? Or is this specific to -symbol-ordering-file and -order_file?

It wasn't broken in general, but it's needed to get `-order_file` working correctly.

> Also, it looks like some of the main changes here by the mangler for objective C are to strip the "\01" prefix. Is this different than the support to remove the '\1' mangling escape in getGlobalIdentifier()?

I think both remove the `\01` prefix. But the mangler might also prepend `_`/`l_` depending on the linkage, and `getGlobalIdentifier()` does not do this.

Reply:
>I think we can only generate the linkage name at compile time because we need to know the linkage, which llvm-profdata doesn't have.

Yep, but my original question is more about why using linkage-name in IRPGOFuncName (early in the pipeline) rather than storing it at a later phase of the pipeline. I read the code and [rfc](https://discourse.llvm.org/t/rfc-temporal-profiling-extension-for-irpgo/68068) again to make sense of why 'linkage-name' is needed in IRPGOFunc names.

> Was PGO broken in general before D156569? Or is this specific to -symbol-ordering-file and -order_file?
> It wasn't broken in general, but it's needed to get -order_file working correctly.

My understanding is that the specific data flow requires linkage-name to be stored in raw profiles. `llvm-profdata order` takes a profile and generates an order file (https://llvm.org/docs/CommandGuide/llvm-profdata.html#order); this file is used by `-symbol-ordering-file` directly.  Correct me if I miss anything.

A similar ELF option to `symbol-ordering-file` is [call-graph-ordering](https://github.com/llvm/llvm-project/blob/005c83380a907becbf5a6b4522fc43652c9536cd/lld/ELF/Driver.cpp#L1675-L1676), but with a different data flow. At a very high level, the edge (<caller, callee, weight> is passed in ELF sections, where linkage-name is already used.

Reply:
> It wasn't broken in general, but it's needed to get -order_file working correctly.

Unfortunately this change broke aspects of ThinLTO ICP, however. Is it possible to change the handling of -order_file in the linker to modify the symbol names appropriately?

> To avoid subtle issues when linkage-name is different from mangled names,,I'm wondering if it warrants a change to use linkage-names (as opposed to mangled name) in GlobalValue::getGlobalIdentifier in this PR.

If the -order_name handling cannot be fixed as I suggested above, then yes, we need some solution to ensure that the hashed symbol names are consistent across PGO and ThinLTO.

Reply:
I think it makes more sense to use linkage-names for IRPGO, `-order_file`, and ThinLTO. `-order_file` is used in the linker when it only knows linkage-names, so I don't think it makes sense to feed it mangled names.

Reply:
>  I think it makes more sense to use linkage-names for IRPGO, -order_file, and ThinLTO. -order_file is used in the linker when it only knows linkage-names, so I don't think it makes sense to feed it mangled names.

Thanks for the input.  (not to bikeshed but) I think for the purpose of computing global identifier, unique names should suffice. Clang-FE mangled names are unique so sounds fine.

Using linkage-name would be a non-trivial change, given the static `getGlobalIdentifier` takes a stringified name currently, and using linkage-name means requiring compatible change in each callsite (e.g., if the caller context doesn't have `GlobalValue` but just stringified names in the bitcode, make sure linkage-name exists in the bitcode, [this](https://github.com/llvm/llvm-project/blob/d6fbd96e5eaf3e8acbf1b43dce7a311352907567/llvm/lib/Bitcode/Reader/BitcodeReader.cpp#L6763) seems one example). An alternative fixup is to store the MD5 of `[filename;]linkage-name` (this is what's currently stored in this [field](https://github.com/llvm/llvm-project/blob/0f45e45847a5f2969b8021c787a566531fc96473/compiler-rt/include/profile/InstrProfData.inc#L72-L74)) and the MD5 of `[filename:]mangled-name` (the original hash before D156569) in the raw profiles, so `llvm-profdata` could choose properly (former for ordering and latter for ICP)

Would you mind if I create a Github issue to track how to fix other potential cases and assign it to you ? This PR would solve the colon and semicolon difference.

Reply:
Thanks for giving the thumb-up Ellis! Just created https://github.com/llvm/llvm-project/issues/74565.